src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs

```diff
@@ -705,12 +725,174 @@ async Task MergeUnversioned(List<ImportResource> inputs, bool keepLastUpdated, b
                     await Merge(inputNoConflict.OrderBy(_ => _.ResourceWrapper.ResourceId).ThenByDescending(_ => int.Parse(_.ResourceWrapper.Version)), keepLastUpdated, useReplicasForReads);
                     loaded.AddRange(inputNoConflict);
                 }
-            }
 
-            async Task Merge(IEnumerable<ImportResource> resources, bool keepLastUpdated, bool useReplicasForReads)
-            {
-                var input = resources.Select(_ => new ResourceWrapperOperation(_.ResourceWrapper, true, true, null, requireETagOnUpdate: false, keepVersion: _.KeepVersion, bundleResourceContext: null)).ToList();
-                await MergeInternalAsync(input, keepLastUpdated, true, false, useReplicasForReads, eventualConsistency, cancellationToken);
+                // Validate if there are any constraint violations
+                (List<ImportResource> ValidResources, List<ImportResource> Conflicts) ValidateConstraintViolations(IReadOnlyList<ImportResource> resources)
+                {
+                    var validResources = new List<ImportResource>();
+                    var conflicts = new List<ImportResource>();
+
+                    var tokenSearchParamRowGenerator = new TokenSearchParamListRowGenerator(_model, _searchParameterTypeMap);
+                    var tokenTokenCompositeSearchParamRowGenerator = new TokenTokenCompositeSearchParamListRowGenerator(_model, tokenSearchParamRowGenerator, _searchParameterTypeMap);
+                    var tokenStringCompositeSearchParamListRowGenerator = new TokenStringCompositeSearchParamListRowGenerator(_model, tokenSearchParamRowGenerator, new StringSearchParamListRowGenerator(_model, _searchParameterTypeMap), _searchParameterTypeMap);
+                    var tokenQuantityCompositeSearchParamListRowGenerator = new TokenQuantityCompositeSearchParamListRowGenerator(_model, tokenSearchParamRowGenerator, new QuantitySearchParamListRowGenerator(_model, _searchParameterTypeMap), _searchParameterTypeMap);
+                    var tokenNumberNumberCompositeSearchParamListRowGenerator = new TokenNumberNumberCompositeSearchParamListRowGenerator(_model, tokenSearchParamRowGenerator, new NumberSearchParamListRowGenerator(_model, _searchParameterTypeMap), _searchParameterTypeMap);
+                    var tokenDateTimeCompositeSearchParamListRowGenerator = new TokenDateTimeCompositeSearchParamListRowGenerator(_model, tokenSearchParamRowGenerator, new DateTimeSearchParamListRowGenerator(_model, _searchParameterTypeMap), _searchParameterTypeMap);
+                    var referenceTokenCompositeSearchParamListRowGenerator = new ReferenceTokenCompositeSearchParamListRowGenerator(_model, new ReferenceSearchParamListRowGenerator(_model, _searchParameterTypeMap), tokenSearchParamRowGenerator, _searchParameterTypeMap);
+
+                    // Traverse through the resources to check which resources are causing constraint violation
+                    foreach (var resource in resources)
```

Comment:
If I read it correctly this code will call Merge for each of input resources. This will have devastating effect on performance.
I agree that it is better provide info on resource level, but not with negative affect on perf.
We should either do "constraint" checks in C#, or just say that batch staring with this resource and ending with this failed with constraint violation, raise bad request, and let customer figure it out.